src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,23 +59,41 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = java.nio.file.Files.createTempDirectory("testngXmlPathInJar-").toFile();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                String jeName = je.getName();
+                if (Parser.canParse(jeName.toLowerCase())){ 
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
+                    if (matchesXmlPathInJar(je)) {
+                        suitePath = copyFile.toString();
                     }
-                    return true;
                 } else if (isJavaClass(je)) {
                     classes.add(constructClassName(je));
                 }
             }
+            if (Strings.isNullOrEmpty(suitePath)) {
+                return foundTestngXml;
+            }
+            Collection<XmlSuite> parsedSuites = Parser.parse(suitePath,processor); 
+            for (XmlSuite suite : parsedSuites) {
+                // If test names were specified, only run these test names
+                if (testNames != null) {
+                  XmlSuiteUtils xmlSuiteUtils = new XmlSuiteUtils();
```

Comment:
Please, find a better class name.

Reply:
Sorry, I didn't catch your mean. You mean rename  `xmlSuiteUtils` or  `XmlSuiteUtils` to a better name?

Reply:
` XmlSuiteUtils` first, and the var name will follow

Reply:
Do you or @krmahadevan have any suggestions ?

Reply:
@eyehwan - I didn't understand your question.

Reply:
I mean if @juherr  thought XmlSuiteUtils is not good name, I would like get both of your suggestions of new name. 
Becaues for me, XmlSuiteUtils is fine...

Reply:
@juherr - Do you want to have `XmlSuiteUtils` renamed ? 
@eyehwan - I dont have any specific suggestion. It was me who extracted out that class as part of a refactoring :)

Reply:
Then @juherr  how about your suggestion? :)

Reply:
@krmahadevan Yes. In my opinion "*Utils" is a pattern marker which means: "a class with static methods, without a state".
As it is now an object, the name should reference the purpose.

Reply:
@juherr  @krmahadevan How about `XmlSuiteHelper`?

Reply:
Or keep `XmlSuiteUtils` unchanged, and only put `validateIfSuitesContainDuplicateTests` and `validateIfSuitesContainDuplicateTests` two method in this class. 
And move all the other functionalities which related to testnames to a new class called "testNamesHelper"?
Again, @krmahadevan @juherr  How do you think about it?

Reply:
"Helper" doesn't explain more the goal than "Utils" or "Things" ;)

Reply:
but at least Helper doesn't have limitation on non-state right? :)
I just push for a new review. If you have new idea, you can tell me.

Reply:
@juherr - Yes your statement on the usage of `Utils`/`Helper`/`Assistant` makes sense. When I created this class it didn't have state. 

@eyehwan - `TestNamesHelper` still suggests the same thing. A class that houses only static utility methods and holds no state. But that's not the case here. Moreover, the implementation for `TestNamesHelper` expects the caller to invoke methods in some order, which is not very intuitive. [ Am referring the code here https://github.com/cbeust/testng/pull/1635/files#diff-c89ea27c996c668601a33f887b5675f2R303 ]

You should instead do the following

```java
TestNamesMatcher matcher = new TestNamesMatcher(s, m_testnames);
matcher.getMissMatchedTestNames();
matcher.getCloneSuites();
```